File: bitcamp-web-04/src/main/java/bitcamp/pms/dao/MemberDao.java
package bitcamp.pms.dao;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.util.ArrayList;
import java.util.List;

import bitcamp.pms.domain.Member;

public class MemberDao {
	
	static {
		try {
			Class.forName("com.mysql.jdbc.Driver");
		} catch (ClassNotFoundException e) {
			e.printStackTrace();
		}
	}
	
	String jdbcUrl;
	String username;
	String password;

	
	
	public MemberDao(
			String jdbcurl, String username, String password) {
		this.jdbcUrl = jdbcurl;
		this.username = username;
		this.password = password;
	}
	
	
	@SuppressWarnings("unused")
	public List<Member> selectList() throws Exception {
		
        try (
            Connection con = DriverManager.getConnection(
        		jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "select mid, email from pms2_member");
            ResultSet rs = stmt.executeQuery();) {
            
            ArrayList<Member> list = new ArrayList<>();
            while (rs.next()) {
                Member member = new Member();
                member.setId(rs.getString("mid"));
                member.setEmail(rs.getString("email"));
                list.add(member);
            }
            return list;
        }
    }
	
	
	
	@SuppressWarnings("unused")
	public Member selectOne(String id) throws Exception {
		
        try (
		Connection con = DriverManager.getConnection(
				jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement("select mid, email from pms2_member where mid=?");
            ) {
    		stmt.setString(1, id);
    		ResultSet rs = stmt.executeQuery();
    		if (!rs.next()) {
    			return null;
    		} else {
    			Member member = new Member();
                member.setId(rs.getString("mid"));
                member.setEmail(rs.getString("email"));
    			return member;
    		}
        }
	}
	
	
	@SuppressWarnings("unused")
	public int update(Member member) throws Exception {
		
        try (
        	Connection con = DriverManager.getConnection(
        			jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "update pms2_member set email=?, pwd=password(?) where mid=?");) {
            
            stmt.setString(1, member.getEmail());
            stmt.setString(2, member.getPassword());
            stmt.setString(3, member.getId());
            
            return stmt.executeUpdate();
        }
	}
	
	
	
	
	@SuppressWarnings("unused")
	public int delete(String id) throws Exception {
		
        try (
        	Connection con = DriverManager.getConnection(
        			jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "delete from pms2_member where mid=?");) {
            
            stmt.setString(1, id);
            
            return stmt.executeUpdate();
        } 
	}
	
	
	
	@SuppressWarnings("unused")
	public void insert(Member member) throws Exception {
		
        try (
        	Connection con = DriverManager.getConnection(
        			jdbcUrl, username, password);
            PreparedStatement stmt = con.prepareStatement(
                "insert into pms2_member(mid,email,pwd) values(?,?,password(?))");) {
            
            stmt.setString(1, member.getId());
            stmt.setString(2, member.getEmail());
            stmt.setString(3, member.getPassword());
        
            stmt.executeUpdate();
        }
	}
}

File: bitcamp-nodejs/src/step05/memberdao.js
// DAO 모듈 만들기

var pool;

exports.setConnectionPool = (connectionPool) => {
    pool = connectionPool;
};

exports.list = (pageNo = 1, pageSize = 3, handler) => {

    var startIndex = (pageNo - 1) * pageSize;
    //this를 사용하여 이 객체의 pool이란걸 명시해야한다.
    //지금은 클로저 pool을 사용하므로 this를 뺐다.
    pool.query(
        'select * from pms2_member limit ?, ?',
        [startIndex, pageSize],
        function(err, results){
            handler(err, results);
    });
};

exports.add = (data, handler) => {
    pool.query(
        `insert into pms2_member(mid, email, pwd)
        values(?, ?, password(?))`,
        [data.id, data.email, data.password],
        function(err, result){
            handler(err, result);
    });
};

exports.update = (data, handler) => {
    pool.query(
        `update pms2_member set email=?, pwd=password(?)
        where mid=?`,
        [data.email, data.password, data.id],
        function(err, result){
            handler(err, result);
    });
};

exports.remove = (data, handler) => {
    pool.query(
        `delete from pms2_member
        where mid=?`,
        [data.id],
        function(err, result){
            handler(err, result);
    });
};

exports.hello = (data, handler) => {
    handler(data);
};